### PR TITLE
eth/filters: add global block logs cache

### PR DESCRIPTION
**Pull request by s1na**
This adds a cache for block logs which is shared by all filters. In order to share the cache reference, filters now need to be created through a 'filter system' object. Cache size is user-configurable via `--cache.blocklogs`.

This PR is an alternative to https://github.com/ethereum/go-ethereum/pull/25425.
cc @ryanschneider 

Fixes https://github.com/ethereum/go-ethereum/issues/25421 
Includes #25468, which should be merged first.

### CODE REVIEW

**Comment by ryanschneider** [comment]
lgtm!  Thanks!

**Review by holiman — APPROVED** [review]
LGTM

File: eth/filters/filter.go
Comment (rjl493456442) [inline eth/filters/filter.go:270]: It's weird to add this "duplicated" check here. I would vote to add the check  here https://github.com/ethereum/go-ethereum/pull/25459/files#diff-9886da3412b43831145f62cec6e895eb3613a175b945e5b026543b7463454603R204

Also in `les_apiBackend.GetLogs`, we should also return the error if logs is not found.

```go
func (b *LesApiBackend) GetLogs(ctx context.Context, hash common.Hash) ([][]*types.Log, error) {
	if number := rawdb.ReadHeaderNumber(b.eth.chainDb, hash); number != nil {
		return light.GetBlockLogs(ctx, b.eth.odr, hash, *number)
	}
	return nil, fmt.Errorf("failed to get logs for block #%d", header.Number.Uint64())
}
```

File: internal/ethapi/backend.go
Comment (fjl) [inline internal/ethapi/backend.go:91]: We need to decide about this. Including the interface is not strictly needed here, because we can assert the interface at runtime in cmd/geth. Might be nicer to catch wrong implementations at compile-time though.

**Comment by fjl** [comment]
This is ready for testing now. The cache size is configurable via `--cache.logcache` or in TOML via `Eth.FilterLogCacheSize = 128`, for example.

@holiman Please re-check this, the PR has changed significantly after your initial review.

File: cmd/utils/flags.go
Comment (karalabe) [inline cmd/utils/flags.go:498]: What does "size" mean? Number of blocks, number of txs, number of logs, bytes, percentage of the total cache allowance?

Comment (fjl) [reply cmd/utils/flags.go:498]: It means number of blocks right now. Need to expand this description.

File: cmd/utils/flags.go
Comment (s1na) [inline cmd/utils/flags.go:496]: nit: isnt this a bit repetitive? cache log cache

Comment (fjl) [reply cmd/utils/flags.go:496]: Yeah it's not the best name, I agree. `--cache.log` is too short. Maybe `--cache.blocklogs`?

**Review by holiman — APPROVED** [review]
Still lgtm!

**Comment by ryanschneider** [comment]
@s1na are the cli flag ergonomics the only thing holding up this PR from being accepted now?  It'd be great to get this in the next release for the Merge!  Looks like one more codeowner ✅ is required.

**Comment by ryanschneider** [comment]
@fjl not yet, but I did list the steps to test the performance regression in https://github.com/ethereum/go-ethereum/issues/25421 so anyone else can as well.

**Comment by ryanschneider** [comment]
@s1na wow great job!

```
$ /tmp/hey-linux-amd64 -c 5 -z 40s -t 0 -m POST -T application/json -d '{"jsonrpc":"2.0", "id": 1123123, "method": "eth_getLogs", "params": [{}]}' http://0.0.0.0:8545/

Summary:
  Total:	40.0023 secs
  Slowest:	0.0429 secs
  Fastest:	0.0011 secs
  Average:	0.0027 secs
  Requests/sec:	1852.4451


Response time histogram:
  0.001 [1]	|
  0.005 [67507]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.009 [592]	|
  0.014 [3885]	|■■
  0.018 [1881]	|■
  0.022 [192]	|
  0.026 [21]	|
  0.030 [10]	|
  0.035 [7]	|
  0.039 [3]	|
  0.043 [3]	|


Latency distribution:
  10% in 0.0014 secs
  25% in 0.0015 secs
  50% in 0.0017 secs
  75% in 0.0020 secs
  90% in 0.0030 secs
  95% in 0.0117 secs
  99% in 0.0153 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0000 secs, 0.0011 secs, 0.0429 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0000 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0047 secs
  resp wait:	0.0023 secs, 0.0009 secs, 0.0405 secs
  resp read:	0.0004 secs, 0.0001 secs, 0.0170 secs

Status code distribution:
  [200]	74102 responses
  ```

Very little of the CPU flame graph was spent in `GetLogs` at all now:

image

**Comment by gzliudan** [comment]
Please add #24949 before #25459

**Pull request merged**
